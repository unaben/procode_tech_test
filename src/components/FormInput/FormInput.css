.todo-input {
  display: flex;
  align-items: center;
  justify-content: center;
  border-bottom: 1px solid gray;
  padding-bottom: 1.6em;
  margin-bottom: 1.6rem;
}

.todo-input-item {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  margin-right: 1.6rem;
}

.todo-input-item label {
  font-weight: bold;
  margin-bottom: 0.7rem;
}

.todo-input-item input {
  padding: 0.6em;
  border: none;
  width: 250px;
}

.todo-input-item input:focus {
  outline: 2px solid var(--clr-primary-color);
}

.primary-btn {
  background: var(--clr-primary-color);
  color: white;
  border: none;
  margin-top: 1.6rem;
  padding: 0.7em;
  width: 3.75rem;
  cursor: pointer;
}

.primary-btn:hover {
  background-color: var(--clr-primary-bg);
}

@media only screen and (max-width: 768px) {
  .todo-input {
    flex-direction: column;
  }

  .todo-input-item .description {
    margin-top: 0.3rem;
  }
}
