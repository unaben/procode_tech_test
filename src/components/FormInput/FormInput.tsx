import { useState, FC } from "react";
import { useAppDispatch } from "../../app/hooks";
import { addTodo } from "../../features/todo/TodoSlice";
import { message } from "../../translation/ENT";
import "./FormInput.css";

type IFormInputProps = {
  setToggleTodoBtn: React.Dispatch<React.SetStateAction<boolean>>;
};

const FormInput: FC<IFormInputProps> = ({ setToggleTodoBtn }) => {
  const [inputData, setInputData] = useState({
    title: "",
    description: "",
  });

  const dispatch = useAppDispatch();

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;
    setInputData((prevValue) => {
      return { ...prevValue, [name]: value };
    });
  };

  const canCreate = [inputData.title, inputData.description].every(Boolean);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (canCreate) {
      dispatch(addTodo(inputData));
      setToggleTodoBtn(false);
      setInputData({
        title: "",
        description: "",
      });
    }
  };

  return (
    <form className="todo-input" onSubmit={handleSubmit}>
      <div className="todo-input-item">
        <label>{message.title}</label>
        <input
          value={inputData.title}
          type="text"
          name="title"
          id="title"
          placeholder="Enter todo title..."
          onChange={handleInputChange}
        />
      </div>
      <div className="todo-input-item">
        <label className="description">{message.description}</label>
        <input
          value={inputData.description}
          type="text"
          name="description"
          id="description"
          placeholder="Enter description...."
          onChange={handleInputChange}
        />
      </div>
      <div className="todo-input-item">
        <button type="submit" className="primary-btn">
          {message.add}
        </button>
      </div>
    </form>
  );
};

export default FormInput;
