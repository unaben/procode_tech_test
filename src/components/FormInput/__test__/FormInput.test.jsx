import { render, screen, waitFor } from "@testing-library/react";
import FormInput from "../FormInput";
import { Provider } from "react-redux";
import { store } from "../../../app/store";
import userEvent from "@testing-library/user-event";

describe("FormInput component", () => {
  it("should renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <FormInput setToggleTodoBtn={jest.fn()} />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should renders button correctly", () => {
    render(
      <Provider store={store}>
        <FormInput setToggleTodoBtn={jest.fn()} />
      </Provider>
    );
    const btnEl = screen.getAllByRole("button");
    expect(btnEl).toHaveLength(1);
  });

  it("Should render all input correctly", () => {
    const mockFunction = jest.fn();
    render(
      <Provider store={store}>
        <FormInput setToggleTodoBtn={mockFunction} />
      </Provider>
    );
    const inputEl = screen.getAllByRole("textbox");
    expect(inputEl).toHaveLength(2);
  });

    it("Should render a text when desciption Input is not empty", async () => {
      const mockFunction = jest.fn();
      const user = userEvent.setup();
      render(
        <Provider store={store}>
          <FormInput setToggleTodoBtn={mockFunction} />
        </Provider>
      );
      const descInput = screen.getByPlaceholderText(/Enter description..../i);
      user.type(descInput, "Go today");

      await waitFor(() => {
        expect(descInput).toHaveValue("Go today");
      });
    });

  it("Should render a string when title Input is empty", () => {
    const mockFunction = jest.fn();
    render(
      <Provider store={store}>
        <FormInput setToggleTodoBtn={mockFunction} />
      </Provider>
    );
    const titleInput = screen.getByPlaceholderText(/Enter todo title.../i);
    userEvent.type(titleInput, "");

    expect(titleInput).toHaveValue("");
  });
});
