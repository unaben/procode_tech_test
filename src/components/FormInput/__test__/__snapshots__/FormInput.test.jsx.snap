// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput component should renders correctly 1`] = `
<div>
  <form
    class="todo-input"
  >
    <div
      class="todo-input-item"
    >
      <label>
        Title
      </label>
      <input
        id="title"
        name="title"
        placeholder="Enter todo title..."
        type="text"
        value=""
      />
    </div>
    <div
      class="todo-input-item"
    >
      <label
        class="description"
      >
        Description
      </label>
      <input
        id="description"
        name="description"
        placeholder="Enter description...."
        type="text"
        value=""
      />
    </div>
    <div
      class="todo-input-item"
    >
      <button
        class="primary-btn"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
</div>
`;
